[filebeat][httpjson] - Added bug-fix for duplicate data issue #33213

ShourieG opened the pull request:
## Type of change
- Bug Fix

## What does this PR do?
This PR adds a bug-fix for the issue #33213 by getting rid of the if condition which was the culprit, and also putting in 
safeguards so the operation remains similar as before and does not break any tests. 

**Issue: The cursor data was not being updated consistently when chaining with pagination at the root level was present.
This has now been fixed & tests have been added to support the same.** 

**response.split, response.transforms would not work for the initial root response when pagination & chaining were present together. This also has been fixed with this change.**

## Why is it important?
Fixes a duplicate data bug as reported in the issue #33213

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
~~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## Author's Checklist
- [x] Code quality & syntax



## Related issues
- Closes #33213 #33212

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

ShourieG commented:
@efd6 updated pr with the requested changes.

efd6 commented on x-pack/filebeat/input/httpjson/request.go:
```diff
@@ -347,10 +347,8 @@ func (r *requester) doRequest(stdCtx context.Context, trCtx *transformContext, p
 				return err
 			}
 			// we will only processAndPublishEvents here if chains & root level pagination do not exist, inorder to avoid unnecessary pagination
```
This comment is no longer valid. It would be helpful to explain here instead that the interval is only cleared when we don't expect we are building a chain. This is the root of the issue so the comment will be valuable.

ShourieG replied:
sure will update that

ShourieG replied:
updated the comment

efd6 commented:
Can you construct a test that will fail without this change based on the original issue?

ShourieG commented:
@efd6 I re-evaluated the fix and saw that we need to mess with the clearInterval() logic. The cursor data was not being updated consistently when chaining with pagination was present and have made changes to fix that which also fixes any issues we had with the clear interval call. Pagination is avoided for the 1st root level request in the 1st iteration when chaining is present and this is done in a way such that the **last_response, last_event & first_event** are all properly updated along with any response.transforms or response.split operations. This was not happening initially and it has now been fixed. Added tests to support the same.
